fedimint-sqlite/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+        opts.disable_statement_logging();
+        let db = SqlitePool::connect_with(opts).await?;
+
+        sqlx::query("CREATE TABLE IF NOT EXISTS kv (key BLOB, value BLOB);")
+            .execute(&db)
+            .await
+            .expect("Error while creating the key-value table");
+
+        Ok(SqliteDb(db))
+    }
+}
+
+#[async_trait]
+impl IDatabase for SqliteDb {
+    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction {
+        let sqlite_dbtx = self.0.begin().await.unwrap();
+        let mut tx = DatabaseTransaction::new(SqliteDbTransaction(sqlite_dbtx), decoders);
+        tx.set_tx_savepoint().await;
+        tx
+    }
+}
+
+#[async_trait]
+impl<'a> IDatabaseTransaction<'a> for SqliteDbTransaction<'a> {
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        let val = self.raw_get_bytes(key).await.unwrap();
+        let query_prepared = sqlx::query("INSERT INTO kv (key, value) VALUES (?, ?)")
+            .bind(key)
+            .bind(value);
+        self.0.execute(query_prepared).await?;
+        Ok(val)
+    }
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let query_prepared =
+            sqlx::query("SELECT value FROM kv WHERE key = ? ORDER BY value DESC LIMIT 1").bind(key);
+        self.0
+            .fetch_optional(query_prepared)
+            .await
+            .and_then(|result| {
+                if result.is_some() {
+                    Ok(Some(result.unwrap().get::<Vec<u8>, &str>("value")))
+                } else {
+                    Ok(None)
+                }
+            })
+            .map_err(anyhow::Error::from)
+    }
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let query_prepared =
+            sqlx::query("SELECT rowid, value FROM kv WHERE key = ? ORDER BY value DESC LIMIT 1")
+                .bind(key);
+        let res = self.0.fetch_optional(query_prepared).await;
+        if res.is_ok() {
+            let optional_row = res.unwrap();
+            if optional_row.is_some() {
+                let row = optional_row.unwrap();
+                let rowid = row.get::<i64, &str>("rowid");
+                let value = row.get::<Vec<u8>, &str>("value");
+
+                let query_prepared = sqlx::query("DELETE FROM kv WHERE rowid = ?").bind(rowid);
+                self.0.execute(query_prepared).await?;
+                return Ok(Some(value));
+            }
+        }
+
+        // Didnt find a key to delete
+        Ok(None)
+    }
+
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> PrefixIter<'_> {
+        // Need to add % to support prefix query
+        let mut prefix = key_prefix.to_vec();
+        prefix.push(b'%');
+
+        let query_prepared =
+            sqlx::query("SELECT key, value FROM kv WHERE key LIKE ? ORDER BY value DESC")
+                .bind(prefix);
```

Comment:
Isn't there a problem that we could have a prefix that already contains a wildcard character like '%' (0x25, the `OutgoingContractAccount` key)? Is there maybe another SQL function we can use for binary prefix search? Otherwise we could format to hex and work on hex values with `LIKE`.

Reply:
Yes this is an issue. Off the top of my head I'm not aware of another prefix match operator but I'll investigate.

Reply:
Formatting to hex seems like the best solution here. I have fixed this and added a test.